### PR TITLE
[FIX] sms: send SMS on transfers without getting an error

### PR DESCRIPTION
**Pull request by mamatch**
Steps to reproduce:
- Create a SMS template with the 'Applies to' property set to Transfer (without this the action will never appear)
- Go to Transfer, pick one, open action and try to send a SMS

Issue:
Traceback

Cause:
When sending the SMS we try to modify the 'mobile' attribute of  'stock.picking' but it doesn't exist.

A test already exists for this case on https://github.com/odoo/odoo/blob/6b7f3f6ec40eb819dd1d94a94722610a63d5697a/addons/test_mail_sms/tests/test_sms_composer.py#L236

opw-3286153

### CODE REVIEW

File: addons/sms/wizard/sms_composer.py [hunk 1]
```diff
@@ -210,7 +210,7 @@ def _action_send_sms_comment_single(self, records=None):
         # on the numbers in the database.
         records = records if records is not None else self._get_records()
         records.ensure_one()
-        if not self.number_field_name:
+        if not self.number_field_name or self.number_field_name not in records:
```
Comment (thiryr) [inline addons/sms/wizard/sms_composer.py:213]: We can keep the check since we already have fallbacks anyway but `Stock Picking` defines `_sms_get_number_fields`, which says it should have a mobile and/or phone field to pick from. So there's an issue there too.

It should defined `_sms_get_partner_fields` instead so we can fallback on the partner instead. Such that this condition would not be reached.

**Review by thiryr — COMMENTED** [review]
I think this should mostly be fixed in `stock_sms` but we can keep this check to avoid tracebacks in the future anyway :)

**Review by thiryr — COMMENTED** [review]
A short test for the fix in `sms` would be neat too :)

Otherwise lgtm functionally

**Comment by mamatch** [comment]
Hi @thiryr , I may be wrong but I think there is already a test which covers that case, https://github.com/odoo/odoo/blob/59f671febe816e1bba3f0eda6059feca45e2bdc1/addons/test_mail_sms/tests/test_sms_composer.py#L123

with a test record model similar to the one we have for ``stock.picking`` in ``stock_sms`` https://github.com/odoo/odoo/blob/784c37b0c0fb1764a214488f69e42836af98ff3f/addons/test_mail_sms/models/test_mail_sms_models.py#L117

**Comment by thiryr** [comment]
hey @mamatch 

Indeed, you/we also added a similar test a while back for a similar issue: `test_composer_sending_with_no_number_field`
But the problem here is different since we do specify a number field, it's just wrong.

I'd say this should be a separate test called `_invalid_field` instead of `nofield`, instead of a subtest. Because one is intended use, the other is us correcting an incorrect use. But you can reuse the logic :)

**Comment by mamatch** [comment]
@thiryr done :slightly_smiling_face:

File: addons/test_mail_sms/tests/test_sms_composer.py [hunk 2]
```diff
@@ -133,6 +133,22 @@ def test_composer_comment_nofield(self):
             sms_composer._action_send_sms()
         self.assertSMSNotification([{'number': self.random_numbers_san[0]}], self._test_body)
 
+    def test_composer_comment_invalid_field(self):
```
Comment (thiryr) [inline addons/test_mail_sms/tests/test_sms_composer.py:136]: Tests should be ordered alphabetically so that they match the order of execution

File: addons/test_mail_sms/tests/test_sms_composer.py [hunk 3]
```diff
@@ -133,6 +133,22 @@ def test_composer_comment_nofield(self):
...
+        sms_composer = self.env['sms.composer'].create({
+            'body': self._test_body,
+            'number_field_name': self.random_numbers_san[0],
+            'res_id': test_record.id,
+            'res_model': 'mail.test.sms.partner'
+        })
+        with self.mockSMSGateway():
+            sms_composer._action_send_sms()
+        self.assertSMSNotification([{'partner': test_record.customer_id}], self._test_body)
+
```
Comment (thiryr) [inline addons/test_mail_sms/tests/test_sms_composer.py:151]: It doesn't seem like the test fails prior to your fix. Always make sure that's the case, it also helps make sure your changes make sense and are valid.

You need to either set `number_field_name` in the composer manually to some field that does not exist (and check that it does not exist, for completeness)

Or patch `_sms_get_number_fields` for `mail.test.sms.partner` so that it returns a non-existing field, which matches the flow better but requires patching.

Your choice.

File: addons/stock_sms/models/stock_picking.py [hunk 4]
```diff
@@ -46,9 +46,12 @@ def _action_generate_warn_sms_wizard(self):
...
     def _sms_get_number_fields(self):
-        """ This method returns the fields to use to find the number to use to
-        send an SMS on a record. """
-        return ['mobile', 'phone']
+        """ No phone or mobile field is available on ``stock.picking`` model. Instead SMS will
+        fallback on partner-based computation using ``_sms_get_partner_fields``. """
+        return []
+
+    def _sms_get_partner_fields(self):
+        return ['partner_id']
```
Comment (thiryr) [inline addons/stock_sms/models/stock_picking.py:54]: This should be a different commit for `stock_sms` as it's a different fix.

**Review by thiryr — COMMENTED** [review]
Hey :wave: 
Some more comments

File: addons/stock_sms/models/stock_picking.py (hunk 4 above)
Comment (mamatch) [reply addons/stock_sms/models/stock_picking.py:54]: Hi @thiryr , a different commit in the same PR, or do I need to create a new one ?

Comment (thiryr) [reply addons/stock_sms/models/stock_picking.py:54]: Just in the same PR, linking the same task :)

It's just so we can clearly separate and justify each change

Comment (mamatch) [reply addons/stock_sms/models/stock_picking.py:54]: Done, hope the commit message is clear enough :)

Comment (thiryr) [reply addons/stock_sms/models/stock_picking.py:54]: Makes sense to me :D

Although note I think the guidelines are normally 50 characters for the title and 72 for the commit lines, including `[TAG] module_name:`

File: addons/test_mail_sms/tests/test_sms_composer.py (hunk 3 above)
Comment (thiryr) [reply addons/test_mail_sms/tests/test_sms_composer.py:152]: Not critical, but can you also check that phone_nbr isn't a field with just `assertNotIn(sms_composer._fields, 'phone_nbr')`?

One never knows..

File: addons/test_mail_sms/tests/test_sms_composer.py (hunk 2 above)
Comment (thiryr) [reply addons/test_mail_sms/tests/test_sms_composer.py:136]: Mm, still not ordered alphabetically

**Review by thiryr — COMMENTED** [review]
Hey
Just a few minor tweaks

File: addons/stock_sms/models/stock_picking.py (hunk 4 above)
Comment (mamatch) [reply addons/stock_sms/models/stock_picking.py:54]: Tried to reduce it the most :)

File: addons/test_mail_sms/tests/test_sms_composer.py (hunk 3 above)
Comment (mamatch) [reply addons/test_mail_sms/tests/test_sms_composer.py]: I had to use ``','.join(test_record._fields)`` because ``assertNotIn()`` waits for a string on the left when the right is also a  string

File: addons/test_mail_sms/tests/test_sms_composer.py [hunk 5]
```diff
@@ -120,6 +120,25 @@ def test_composer_comment_field_w_template(self):
...
+        })
+        sms_composer = self.env['sms.composer'].create({
+            'body': self._test_body,
+            'number_field_name': 'phone_nbr',
+            'recipient_single_number_itf': self.random_numbers_san[0],
+            'res_id': test_record.id,
+            'res_model': 'mail.test.sms.partner'
+        })
+
+        self.assertNotIn(','.join(test_record._fields), 'phone_nbr')
```
Comment (thiryr) [inline addons/test_mail_sms/tests/test_sms_composer.py:137]: No need to join. Unlike js you can check that a string is in a list directly, and it's actually more efficient (and more intuitive)

**Review by thiryr — APPROVED** [review]
Minor comments but otherwise lgtm. Thanks for the fix!

File: addons/test_mail_sms/tests/test_sms_composer.py (hunk 5 above)
Comment (mamatch) [reply addons/test_mail_sms/tests/test_sms_composer.py:137]: It doesn't work because the ``_fields`` returns a dict and not a list :/

**Comment by kitan191** [comment]
@robodoo delegate+

**Comment by mamatch** [comment]
@robodoo r+

**Comment by mamatch** [comment]
@robodoo rebase-ff

**Comment by mamatch** [comment]
@robodoo r+
